### PR TITLE
Tiltfile should tolerate creds set in base 64 encoding

### LINKED ISSUE #1573: Tiltfile should tolerate creds set in base 64 encoding
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

We should reorder the tiltfile so that auth validation (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/Tiltfile#L97) happens after decoding B64 variables (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/Tiltfile#L218-L221) so that B64 values are included when checking if the right creds are set.

Suggestion is to move

```
    substitutions = settings.get("kustomize_substitutions", {})
    for key in keys:
        if key[-4:] == "_B64":
            os.environ[key[:-4]] = base64_decode(os.environ[key])
```

to the beginning on `validate_auth`.           

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sayantani11**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1573 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: Tiltfile
Comment (CecileRobertMichon) [inline Tiltfile:218]: we need to leave this line since substitutions is used later on line 223

`Tiltfile:223:8: undefined: substitutions`

File: Tiltfile
Comment (CecileRobertMichon) [inline Tiltfile:99]: I realize now this won't work because it's looking at the keys instead of the substitutions. Here's what this should be:

```
def validate_auth():
    substitutions = settings.get("kustomize_substitutions", {})
    os.environ.update(substitutions)
    for sub in substitutions:
        if sub[-4:] == "_B64":
            os.environ[sub[:-4]] = base64_decode(os.environ[sub])
    missing = [k for k in keys if not os.environ.get(k)]
    if missing:
        fail("missing kustomize_substitutions keys {} in tilt-setting.json".format(missing))
```

This will set all the variables in substitutions as environment variables and then use those to set the decoded equivalent variable if the substitution had a _B64 suffix.

Since we're doing `os.environ.update(substitutions)` early now, you can delete the line in `dev capz():` that does `os.environ.update(substitutions)` and the one before (lines 136-137).

Comment (CecileRobertMichon) [reply Tiltfile:99]: we also need to change line 193 to

```
    os.putenv('AZURE_CLIENT_SECRET_B64', base64_encode(os.environ.get("AZURE_CLIENT_SECRET")))
```

and remove the `substitutions = settings.get("kustomize_substitutions", {})` line above (line 192)

So it gets the secret directly from the environment variables

Comment (CecileRobertMichon) [reply Tiltfile:99]: Finally, let's add `echo_off=True` to all the local() functions that might print secrets so we don't output those in logs (I spotted 3 in the file, in `base64_encode`, `base64_decode` and `create_identity_secret`). For example:

```
    encode_blob = local("echo '{}' | tr -d '\n' | base64 - | tr -d '\n'".format(to_encode), quiet=True, echo_off=True)
```

**Review by CecileRobertMichon — CHANGES_REQUESTED** [review]
@sayantani11 I pointed you in the wrong direction before, apologies for the confusion. I've added comments on what needs to be changed, let me know if you have questions.

Again, you can test this by running `make tilt-up` locally after following https://capz.sigs.k8s.io/developers/development.html#tilt-for-dev-in-capz to set up your tilt-settings.json file

**Comment by sayantani11** [comment]
@CecileRobertMichon I was also thinking that, because although I was able to open it in local host, it showed tiltfile error.
But I thought, it was just my device :)

**Comment by sayantani11** [comment]
``` <builtin>: in read_json
Error: error parsing JSON from tilt-settings.json: invalid character '}' looking for beginning of object key string ```
@CecileRobertMichon the local host still has this error.
Does tilt-setting.json need change? or the problem is from my side?

**Comment by CecileRobertMichon** [comment]
@sayantani11 that error looks to be a formatting error in your tilt-settings.json

Did you follow https://capz.sigs.k8s.io/developers/development.html#tilt-for-dev-in-capz to set it up? Do you have environment variables set for AZURE_SUBSCRIPTION_ID, etc. ? If not, you could set them to fake values for now just so you can test tilt up.

**Comment by sayantani11** [comment]
@CecileRobertMichon I did but let me look into it again.

**Comment by sayantani11** [comment]
@CecileRobertMichon Guess I am making some mistake in entering the environment variable.
So suppose,
``` "AZURE_SUBSCRIPTION_ID_B64": "$(echo "${AZURE_SUBSCRIPTION_ID}" | tr -d '\n' | base64 | tr -d '\n')" ```
I use the fake ID, how am I filling this up?

**Comment by CecileRobertMichon** [comment]
@sayantani11 try:

```
{
  "kustomize_substitutions": {
      "AZURE_SUBSCRIPTION_ID_B64": "ZmFrZXZhbHVlCg==",
      "AZURE_TENANT_ID_B64": "ZmFrZXZhbHVlCg==",
      "AZURE_CLIENT_SECRET_B64": "ZmFrZXZhbHVlCg==",
      "AZURE_CLIENT_ID_B64": "ZmFrZXZhbHVlCg==",
  }
}
```

`ZmFrZXZhbHVlCg==` is "fakevalue" in base 64.

**Comment by sayantani11** [comment]
@CecileRobertMichon Yeah I actually did it but forgot to push. Doing it now.

**Comment by sayantani11** [comment]
@CecileRobertMichon It came to notice that  https://capz.sigs.k8s.io/developers/development.html#tilt-for-dev-in-capz doesn't mention installation of Tilt as a Makefile dependency. So it would be better if it's added, as I was getting the error at first.

**Comment by CecileRobertMichon** [comment]
>  It came to notice that https://capz.sigs.k8s.io/developers/development.html#tilt-for-dev-in-capz doesn't mention installation of Tilt as a Makefile dependency. So it would be better if it's added, as I was getting the error at first.

Let's do this as a follow-up. The rest of the changes look good, thanks again for your patience here!

/lgtm
/assign @nader-ziada

**Comment by CecileRobertMichon** [comment]
@sayantani11 before this merges, can you please squash the commits?

/hold
/ok-to-test

**Pull request merged**
